fin: add of_le, cast and swap_add

tobiasgrosser opened the pull request:
TO CONTRIBUTORS:

Make sure you have:

 * [x] reviewed and applied the coding style: [coding](./docs/style.md), [naming](./docs/naming.md)
 * [x] make sure definitions and lemmas are put in the right files
 * [x] make sure definitions and lemmas are not redundant

For reviewers: [code review check list](./docs/code-review.md)

spl commented on data/fin.lean:
```diff
@@ -100,3 +100,6 @@ def lower_left {n m} (i : fin (n + m)) (h : i.val < n) : fin n :=
 def lower_right {n m} (i : fin (n + m)) (h: i.val >= n) : fin m :=
 ⟨i.val - n, by simp [nat.sub_lt_right_iff_lt_add h, i.is_lt]⟩
+
+def swap_sum {n m}  (i: fin (n + m)) : fin (m + n) :=
```
I'm not sure about the name. Perhaps `swap_add` or `add_comm`? I don't think `sum` is generally used in mathlib to refer to `+`.

tobiasgrosser replied:
I like swap_add.

tobiasgrosser committed:
fin: add raise_le, cast, swap_add (and rename raise -> raise_one)

fin.raise_le allows a conversion from fin n to fin m given n <= m.

fin.cast allows a conversion from fin n to fin m given n = m.

fin.swap_add swaps operands when size is given as addition.

Also express fin.raise_nat as fin.raise_le.

Rename fin.raise to fin.raise_one express it as fin.raise_le.

cipher1024 commented on data/fin.lean:
```diff
@@ -100,3 +100,9 @@ def lower_left {n m} (i : fin (n + m)) (h : i.val < n) : fin n :=
+def fin.swap_add {n m}  (i: fin (n + m)) : fin (m + n) :=
+⟨i.val, by simp [i.is_lt]⟩
+
+def fin.cast {n m}  (i: fin n) (h: n = m) : fin m :=
```
It might be just me but I prefer the opposite order of `i` and `h`. This way, if you have a `xs : list (fin n)`, you can do `xs.map (fin.cast h)` directly.

tobiasgrosser replied:
Changed.

tobiasgrosser committed:
fin: use namespace prefix for previously added function

For some reason the later functions/lemmas in fin.lean use explicit name
spaces rather then an open/close namespace. This change ensures all new
functions have a 'fin' prefix.

In a later change one might consider to expand the namespace to avoid
the need for 'fin.' prefixes.

tobiasgrosser committed:
data/fin: remove eq_of_lt_succ_of_not_lt

This function does not belong in data/fin and has no user outside of
data/fin and the last use of it in data/fin got removed in
(refactor(data/equiv,encodable): refactor/simplify proofs, Tue Sep 19
02:55:14 2017 -0400).

tobiasgrosser committed:
data/fin: Move everything in 'fin' namespace

spl commented on data/fin.lean:
```diff
@@ -40,63 +37,73 @@ instance {n : ℕ} : decidable_linear_order (fin n) :=
 def lower_right {n m} (i : fin (n + m)) (h: i.val >= n) : fin m :=
 ⟨i.val - n, by simp [nat.sub_lt_right_iff_lt_add h, i.is_lt]⟩
+
+def raise_le {n m} (h: n ≤ m) (i: fin n) : fin m :=
```
Style remarks:
1. Personally, I prefer alphabetical order for variables unless there is a reasonable (e.g. semantic) alternative. It just makes it easier to follow. (Also, in this case the alphabetical order matches the comparison.) So, `{m n}`, `m ≤ n`, etc.
2. Single space on either side of `:`.

spl commented on data/fin.lean:
```diff
@@ -40,63 +37,73 @@ instance {n : ℕ} : decidable_linear_order (fin n) :=
+def raise_le {n m} (h: n ≤ m) (i: fin n) : fin m :=
+⟨i.val, lt_of_lt_of_le (i.is_lt) h⟩
+
+def cast {n m} (h: n = m) (i: fin n) : fin m :=
```
Apply style changes as mentioned above.

spl commented on data/fin.lean:
```diff
@@ -40,63 +37,73 @@ instance {n : ℕ} : decidable_linear_order (fin n) :=
+⟨i.val, lt_of_lt_of_le (i.is_lt) h⟩
+
+def cast {n m} (h: n = m) (i: fin n) : fin m :=
+fin.raise_le (by simp[h]) i
```
Single space after `simp`: `simp [h]`.

tobiasgrosser replied:
Done.

spl reviewed (changes requested):
Other than a few minor style remarks, it is 👍. I don't have a strong opinion on `fin.cast`. It seems like a useful thing to have.
